### PR TITLE
Optimise QQ memory use

### PR DESCRIPTION
**Pull request by kjnilsson**
Take fewer release cursor snapshots points as the message backlog grows.

Also introduces a compacted form of the internal message header map
where initially it is only an integer representing the size of the
message body. Later when additional keys need to be added it is expanded
into a full map. This avoid creating and holding many individial maps
with just a size element.

[#169064158]

### CODE REVIEW

File: src/rabbit_fifo.erl [hunk 1]
```diff
@@ -974,26 +976,48 @@ append_to_master_index(RaftIdx,
     State#?MODULE{ra_indexes = Indexes}.
 
 
+incr_enqueue_count(#?MODULE{cfg = #cfg{release_cursor_interval = C} = Cfg}
+                   = State0)
+  when is_integer(C) ->
+    %% conversion to new release cursor interval format
+    State = State0#?MODULE{cfg = Cfg#cfg{release_cursor_interval = {C, C}}},
+    incr_enqueue_count(State);
```
Comment (dumbbell) [inline src/rabbit_fifo.erl]: This function clause could go after the next one:
* the most common clause appears first, which could be more readable for developpers
* this is a small optimization (the most common clause is evaluated first)

Comment (dumbbell) [reply src/rabbit_fifo.erl]: Likewise for other function clauses which do the conversion.

Comment (kjnilsson) [reply src/rabbit_fifo.erl]: Hmm yes that may actually apply to pretty much all "overloaded" functions. I'll update.

Comment (kjnilsson) [reply src/rabbit_fifo.erl]: That said the compiler might do this optimisation for us.

**Comment by michaelklishin** [comment]
This was backported to `v3.8.x`.

**Pull request merged**